### ISSUE TITLE
Use Django rest framework client instead of django client when testing

### ISSUE THREAD

**Issue by silentninja** [centerofci/mathesar#813]
## Problem
The test cases are using Django client which is much suited for emulating a browser, so the default request body is priority content type is multipart instead of json which is what we are using mostly on the frontend

## Proposed solution
DRF [test client](https://www.django-rest-framework.org/api-guide/testing/) should be used instead of django client in order to emulate an API call rather than a browser form, this avoids unnecessary json dumps when calling an API during testing

**Comment by ghislaineguerin** [comment]
@silentninja Thanks for opening this issue. Is this ready to be worked on, or do you need to include additional details?

**Comment by silentninja** [comment]
This is ready to be worked on but have to check if there is any feedback before marking it ready

**Comment by kgodey** [comment]
@silentninja if you'd like feedback on an issue, the best way to get it is tag the people you need feedback from in a issue comment or bring the issue up on Matrix.

I think it's a good idea to switch to the DRF test client.

**Comment by andruwwwka** [comment]
Hi! 
Can I help you with this issue?

**Comment by silentninja** [comment]
Hey! @andruwwwka Sure, I have assigned it to you.

**Comment by andruwwwka** [comment]
@silentninja thx!

**Comment by silentninja** [comment]
There is a slight change in the issue, it seems like we are already using [DRF test client](https://github.com/centerofci/mathesar/blob/e55088bf7100c9ca4347c2a251ac0e424f71573b/mathesar/tests/api/conftest.py#L22). But the test cases are not using correct options for sending json data, let me know if you are okay to work on it I will update the issue

**Comment by silentninja** [comment]
My IDE was pointing to the wrong client fixture, but the seems to be using the correct API client in [contest](https://github.com/centerofci/mathesar/blob/e55088bf7100c9ca4347c2a251ac0e424f71573b/mathesar/tests/api/conftest.py#L22), the problem seems to be with improper settings

**Comment by andruwwwka** [comment]
@silentninja You are right, DRF test client is already being used.
The request format is not specified in the test cases. For example:
```python
def test_column_update_delete_default(column_test_table, client):
    cache.clear()
    expt_default = None
    data = json.dumps({"default": None})
    response = client.patch(
        f"/api/v0/tables/{column_test_table.id}/columns/2/", data=data,
        content_type="application/json"
    )
    assert response.json()["default"] == expt_default
```
It works fine if we set `format` parameter for `post`, `patch`, `put` methods:
```python
def test_column_update_delete_default(column_test_table, client):
    cache.clear()
    expt_default = None
    data = {"default": None}
    response = client.patch(
        f"/api/v0/tables/{column_test_table.id}/columns/2/",
        data=data,
        format="json",
    )
    assert response.json()["default"] == expt_default
```
However, for `get` method if we use nested dict it not works. This is because [get method of APIRequestFactory](https://github.com/encode/django-rest-framework/blob/3.12.4/rest_framework/test.py#L194) use [urlencode from django](https://github.com/django/django/blob/3.1.13/django/utils/http.py#L104) for convert query string. It not support nested dict and convert not-string nested values via str() function. It generates non-serializable strings. In this case dumps is a true way
For example:
```python
def test_record_list_filter(create_table, client):
    table_name = 'NASA Record List Filter'
    table = create_table(table_name)

    filter_list = [
        {'or': [
            {'and': [
                {'field': 'Center', 'op': '==', 'value': 'NASA Ames Research Center'},
                {'field': 'Case Number', 'op': '==', 'value': 'ARC-14048-1'}
            ]},
            {'and': [
                {'field': 'Center', 'op': '==', 'value': 'NASA Kennedy Space Center'},
                {'field': 'Case Number', 'op': '==', 'value': 'KSC-12871'}
            ]}
        ]}
    ]
    json_filter_list = json.dumps(filter_list)

    with patch.object(
        models, "get_records", side_effect=models.get_records
    ) as mock_get:
        response = client.get(
            f'/api/v0/tables/{table.id}/records/?filters={json_filter_list}'
        )
        response_data = response.json()

    assert response.status_code == 200
    assert response_data['count'] == 2
    assert len(response_data['results']) == 2
    assert mock_get.call_args is not None
    assert mock_get.call_args[1]['filters'] == filter_list

```
I can fix the test cases except `get` requests with nested dict

**Comment by silentninja** [comment]
@andruwwwka That's correct, `get` request can only accept strings, so the nested dictionary needs to be converted into a json string and parsed into a json on the backend. This is the reason GraphQL and such use  `post` request

**Comment by andruwwwka** [comment]
@silentninja Is the issue still relevant? Fix test cases without nested dict in `get` requests?

**Comment by silentninja** [comment]
> @silentninja Is the issue still relevant? Fix test cases without nested dict in `get` requests?

Yes the underlying problem is still the same, so I have edited the issue title and description to reflect the problem correctly
